drivers/flash/flash_disk.c

```diff
@@ -0,0 +1,312 @@
+release_device:
+	release_device(dev);
+	return ret;
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void flash_disk_pages_layout(const struct device *dev,
+				    const struct flash_pages_layout **layout, size_t *layout_size)
+{
+	LOG_DBG("%s called", __func__);
+	const struct flash_disk_dev_config *cfg = dev->config;
+
+	*layout = &cfg->flash_pages_layout;
+	*layout_size = 1;
+}
+#endif
+
+static const struct flash_parameters *flash_disk_get_parameters(const struct device *dev)
+{
+	LOG_DBG("%s called", __func__);
+	const struct flash_disk_dev_config *cfg = dev->config;
+
+	return &cfg->flash_parameters;
+}
+
+static const struct flash_driver_api disk_flash_api = {
+	.read = flash_disk_read,
+	.write = flash_disk_write,
+	.erase = flash_disk_erase,
+	.get_parameters = flash_disk_get_parameters,
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = flash_disk_pages_layout,
+#endif
+#if defined(CONFIG_FLASH_JESD216_API)
+	.sfdp_read = NULL,
+	.read_jedec_id = NULL,
+#endif
+#if defined(CONFIG_FLASH_EX_OP_ENABLED)
+	.ex_op = NULL,
+#endif
+};
+
+static void acquire_device(const struct device *dev)
+{
+	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
+		struct flash_disk_dev_data *const data = dev->data;
+
+		k_sem_take(&data->sem, K_FOREVER);
+	}
+}
+
+static void release_device(const struct device *dev)
+{
+	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
+		struct flash_disk_dev_data *const data = dev->data;
+
+		k_sem_give(&data->sem);
+	}
+}
+
+/* clang-format off */
+#define DISK_FLASH_DEFINE(index) \
+	char buf_##index[DT_INST_PROP(index, disk_sector_size)]; \
+	static struct flash_disk_dev_data disk_flash_data_##index = { \
+		.buf = buf_##index, \
+		.bufsize = sizeof(buf_##index), \
+	}; \
+\
+	BUILD_ASSERT(DT_INST_PROP(index, size) > 0, "flash_size needs to be > 0"); \
+	BUILD_ASSERT(DT_INST_PROP(index, write_block_size) > 0, \
+		     "write_block_size needs to be > 0"); \
+	BUILD_ASSERT(DT_INST_PROP(index, erase_value) > 0 && \
+			     DT_INST_PROP(index, erase_value) <= 0xFF, \
+		     "erase_value has to be between 0x00 and 0xFF"); \
+	BUILD_ASSERT(DT_INST_PROP(index, disk_offset) >= 0, "disk_offset needs to be >= 0"); \
+	static const struct flash_disk_dev_config disk_flash_config_##index = { \
+		.flash_size = DT_INST_PROP(index, size), \
+		IF_ENABLED( \
+			CONFIG_FLASH_PAGE_LAYOUT, \
+			(.flash_parameters = \
+				{ \
+					.write_block_size = DT_INST_PROP(index, write_block_size), \
+					.erase_value = DT_INST_PROP(index, erase_value) \
+				}, \
+			) \
+		) \
+		.disk_name = DT_INST_PROP(index, disk_name), \
+		.disk_offset = DT_INST_PROP(index, disk_offset), \
+	}; \
+\
+	BUILD_ASSERT(DT_INST_PROP(index, init_priority) >= 0, "init_priority needs to be >= 0"); \
+	DEVICE_DT_INST_DEFINE(index, &flash_disk_init, NULL, &disk_flash_data_##index, \
+			      &disk_flash_config_##index, \
+			      DT_STRING_TOKEN(DT_DRV_INST(index), init_level), \
+			      DT_INST_PROP(index, init_priority), &disk_flash_api);
+/* clang-format on */
```

Comment:
Is this comment needed?

Reply:
Well it is not really needed but if clang-format is not set to `off` the whole `define` gets formatted in a really weird way.